fsm/observer.go

```diff
@@ -0,0 +1,78 @@
+package fsm
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+// CachedObserver is an observer that caches all states and transitions of
+// the observed state machine.
+type CachedObserver struct {
+	lastNotification    Notification
+	cachedNotifications []Notification
+
+	notificationCond *sync.Cond
+	notificationMx   sync.Mutex
+}
+
+// NewCachedObserver creates a new cached observer.
+func NewCachedObserver() *CachedObserver {
+	observer := &CachedObserver{
+		cachedNotifications: make([]Notification, 0),
+	}
+	observer.notificationCond = sync.NewCond(&observer.notificationMx)
+
+	return observer
+}
+
+// Notify implements the Observer interface.
+func (c *CachedObserver) Notify(notification Notification) {
+	c.notificationMx.Lock()
+	defer c.notificationMx.Unlock()
+
+	c.cachedNotifications = append(c.cachedNotifications, notification)
+	c.lastNotification = notification
+	c.notificationCond.Broadcast()
+}
+
+// GetCachedNotifications returns the cached notifications.
+func (c *CachedObserver) GetCachedNotifications() []Notification {
+	return c.cachedNotifications
```

Comment:
We need to make a copy of the slice since it can be written by `Notify()`. Also would be nice to have a function to clear the cache slice.